Tchap/Managers/Registration/RegistrationService.swift

```diff
@@ -0,0 +1,237 @@
+/*
+ Copyright 2018 New Vector Ltd
+ 
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+ 
+ http://www.apache.org/licenses/LICENSE-2.0
+ 
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+import Foundation
+
+/// `RegistrationService` implementation of `RegistrationServiceType` is used to register a user on Tchap platforms.
+final class RegistrationService: RegistrationServiceType {
+    
+    // MARK: - Constants
+    
+    private enum Constants {
+        static let homeServerSubDomain = "matrix"
+        static let webAppSubDomain = "chat"
+        static let registrationRetryInterval: TimeInterval = 10.0
+    }
+    
+    // MARK: - Properties
+    
+    private let accountManager: MXKAccountManager
+    private let restClient: MXRestClient
+    
+    private var registrationOperation: MXHTTPOperation?
+    private var registrationRetryTimer: Timer?
+    
+    // MARK: - Setup
+    
+    init(accountManager: MXKAccountManager, restClient: MXRestClient) {
+        self.accountManager = accountManager
+        self.restClient = restClient
+    }
+    
+    // MARK: - Public
+    
+    func submitRegistrationEmailVerification(to email: String, completion: @escaping (MXResponse<ThreePIDCredentials>) -> Void) {
+        self.submitRegistrationEmailVerification(to: email, using: self.restClient, completion: completion)
+    }
+    
+    func register(with threePIDCredentials: ThreePIDCredentials, password: String, deviceDisplayName: String, completion: @escaping (MXResponse<String>) -> Void) {
```

Comment:
We have to highlight here that the registration is based on the email identity flow type.
I suggest renaming this method with:
`func register(withEmailIdentity threePIDCredentials: ThreePIDCredentials, password:...`